projects/plugins/jetpack/extensions/store/membership-products/resolvers.js

```diff
@@ -25,88 +25,107 @@ import { getMessageByProductType } from '../../shared/components/product-managem
+	return response;
+};
 
-		if (
-			! response?.products?.length &&
-			! response.should_upgrade_to_access_memberships &&
-			response.connected_account_id &&
-			! selectedProductId
-		) {
-			// Is ready to use and has no product set up yet. Let's create one!
-			await dispatch(
-				saveProduct(
-					{
-						title: getMessageByProductType( 'default new product title', productType ),
-						currency: 'USD',
-						price: 5,
-						interval: '1 month',
-					},
-					productType,
-					setSelectedProductId
-				)
-			);
-			dispatch( setApiState( API_STATE_CONNECTED ) );
-			executionLock.release( lock );
-			return;
-		}
+const hydrateMembershipProductsStoreData = ( response, registry, dispatch ) => {
+	const postId = registry.select( editorStore ).getCurrentPostId();
+
+	dispatch( setConnectUrl( getConnectUrl( postId, response.connect_url ) ) );
+	dispatch( setShouldUpgrade( response.should_upgrade_to_access_memberships ) );
+	dispatch( setSiteSlug( response.site_slug ) );
+	dispatch( setUpgradeUrl( response.upgrade_url ) );
+	dispatch( setProducts( response.products ) );
+};
+
+const createDefaultProduct = async ( productType, setSelectedProductId, dispatch ) => {
+	await dispatch(
+		saveProduct(
+			{
+				title: getMessageByProductType( 'default new product title', productType ),
+				currency: 'USD',
+				price: 5,
+				interval: '1 month',
+			},
+			productType,
+			setSelectedProductId
+		)
+	);
+};
+
+const shouldCreateDefaultProduct = response =>
+	! response.products.length &&
+	! response.should_upgrade_to_access_memberships &&
+	response.connected_account_id;
+
+const setDefaultProductIfNeeded = ( selectedProductId, setSelectedProductId, select ) => {
+	if ( ! selectedProductId ) {
+		const defaultProductId = select.getProductsNoResolver()[ 0 ].id;
+		setSelectedProductId( defaultProductId );
+	}
+};
 
-		if ( response?.products?.length > 0 ) {
-			dispatch( setProducts( response.products ) );
-			if ( ! selectedProductId ) {
-				setSelectedProductId( response.products[ 0 ].id );
-			}
+export const getProducts = (
+	productType = PRODUCT_TYPE_PAYMENT_PLAN,
+	selectedProductId = 0,
+	setSelectedProductId = () => {}
+) => async ( { dispatch, registry, select } ) => {
+	await executionLock.blockExecution( EXECUTION_KEY );
+	if ( hydratedFromAPI ) {
+		setDefaultProductIfNeeded( selectedProductId, setSelectedProductId, select );
+		return;
+	}
+
+	try {
+		const lock = executionLock.acquire( EXECUTION_KEY );
+		const response = await fetchMemberships();
+		hydrateMembershipProductsStoreData( response, registry, dispatch );
+
+		if ( shouldCreateDefaultProduct( response ) ) {
+			// Is ready to use and has no product set up yet. Let's create one!
+			await createDefaultProduct( productType, setSelectedProductId, dispatch );
 		}
 
 		dispatch(
 			setApiState( response.connected_account_id ? API_STATE_CONNECTED : API_STATE_NOTCONNECTED )
 		);
+
+		setDefaultProductIfNeeded( selectedProductId, setSelectedProductId, select );
```

Comment:
This Store should not be responsible for setting the default product that the controls should be using. I'll create a task to refactor this out to the actual Controls and then remove the invocations and definition from this file.